### DISCUSSION TITLE
Total completion token count

### DISCUSSION THREAD

**Discussion by oferavnery** [betalgo/openai#336] [Q&A]
With the new functions defined as part of the request, combined with streaming output, it is becoming very tricky and probably inaccurate to measure total  messages + functions + response tokens. I'm building an app that charges end users by their token usage and I want to be as close as possible to what OpenAI logs and charges. 

The wish is to incorporate a way to easily get to those total response / request token counts both in streaming and standard mode.

Thanks! Awesome progress!

**Comment by oferavnery** [comment]
Anybody else is looking to get total token count including request and response as part of the API functionality?

**Comment by kayhantolga** [comment]
Usage should be available in response messages. Are you looking for something else?
![image](https://github.com/betalgo/openai/assets/1096362/ee948d0d-dc37-4f21-9745-026c5cce64a3)
![image](https://github.com/betalgo/openai/assets/1096362/1b0d98cd-c2a3-458a-b0a3-98f7be94333e)

**Reply by oferavnery** [reply]
I don't think it's available in streaming at all. Also, tokens should be counted for the Request body as well as for the response, including functions and messages, and I don't think there's a good way to do that too.

**Comment by kayhantolga** [comment]
You are correct, the Stream Endpoint does not provide usage information. I believe that `PromptTokens` refers to the request tokens, while `CompletionTokens` refers to the response tokens, which are used to calculate overall usage.

Another option is to use the Tokenizer, but currently, it is not working perfectly.

**Comment by richardtallent-erm** [comment]
Is there a way to access the raw JSON sent and received, so we can use a tokenizer to get a more accurate token count? The prompt text and completion text isn't the whole story, apparently.
